### PR TITLE
Increase Audio DMA Interrupt delay to 200 cycles (Fix assorted Namco games)

### PR DESCRIPTION
**Pull request by phire**
We really need hardware tests for this one. But this will do as a quick fix for [issue 9509](https://bugs.dolphin-emu.org/issues/9509) for 5.0

The fix is slightly fragile, so I'd like all games tested.
Games to Test:

 * [x] Xenoblade Chronicles
 * [x] Pokemon Snap
 * [x] Harvest Moon: Magical Melody
 * [x] Final Fantasy Crystal Chronicles
 * [x] Pokemon Box
 * [x] Endless Ocean 1/2 
 * [x] Pokemon Colosseum Bonus Disc
 * [x] RSQEAF - Wii Ski 
 * [x] STGJAF - Tales of Graces (Namco)
 * [x] R2JJAF - Taiko no Tatsujin (Namco)
 * [x] S2TJAF - Taiko no Tatsujin Wii: Dodoon to 2-Daime! (Namco)
 * [x] S3TJAF - Taiko no Tatsujin Wii: Minna de Party 3-Daime! (Namco)
 * [x] STJJAF - Taiko no Tatsujin Wii: Ketteiban (Namco)
 * [x] S5KJAF - Taiko no Tatsujin Wii: Chou Goukaban (Namco)
 * [x] RQRJAF - The Sky Crawlers: Innocent Aces (Namco)
 * [x] RYKJAF - We Ski & Snowboard (Namco)
 * [x] RQREXF - The Sky Crawlers: Innocent Aces (USA)

---
This change is totally [Reviewable](https://reviewable.io/reviews/dolphin-emu/dolphin/3829)

### CODE REVIEW

**Comment by skidau** [comment]
Pokemon Snap and Final Fantasy Crystal Chronicles have to be tested with this one too.

**Comment by phire** [comment]
And for good measure, anything affected by PR #3800 I know from my previous testing that pokemon box crashes when this very interrupt is delayed by 3600 or more cycles.

**Comment by mbc07** [comment]
Pokémon Colosseum Bonus Disc working fine here with this PR...

**Comment by phire** [comment]
Got a bunch of tests from @jmc4789. Apparently Tales of Graces still doesn't boot.

**Comment by JMC47** [comment]
Oh whoops.  Tales of Graces does work.  Was loading the wrong thing!  Sorry!

**Comment by JMC47** [comment]
I also booted Tales of Symphonia: Dawn of a New World since it's also Namco.  I already tested the GameCube one since the merge and it's fine.

File: Source/Core/Core/HW/DSP.cpp [hunk 1]
```diff
@@ -428,10 +428,17 @@ void RegisterMMIO(MMIO::Mapping* mmio, u32 base)
...
+
 				// We make the samples ready as soon as possible
 				void *address = Memory::GetPointer(g_audioDMA.SourceAddress);
 				AudioCommon::SendAIBuffer((short*)address, g_audioDMA.AudioDMAControl.NumBlocks * 8);
-				CoreTiming::ScheduleEvent(80, et_GenerateDSPInterrupt, INT_AID);
+
+				// TODO: need hardware tests for the timing of this interrupt.
+				// Sky Crawlers crashes at boot if this is scheduled less than 87 cycles in the future.
+				// Other Namco games crash too, see issue 9509. For now we will just push it to 200 cycles
+				CoreTiming::ScheduleEvent(200, et_GenerateDSPInterrupt, INT_AID);
```
Comment (degasus) [inline Source/Core/Core/HW/DSP.cpp]: I guess this event should be triggered based on g_audioDMA.AudioDMAControl.NumBlocks. Set this interrupt once we're able to read in the future doesn't sound nice.

Comment (phire) [inline Source/Core/Core/HW/DSP.cpp]: Nope, NumBlocks shouldn't affect it. This is an interrupt to say that the DMA has started, After NumBlocks * 120,000 cycles the interrupt fires as the DMA restarts.

More precisely it says the values of the registers have been latched into the DMA and you may now write new values which will be automatically latched when the DMA finishes.

**Comment by JMC47** [comment]
Round 3 of the timing problems has been defeated.  But for how long?

**Comment by JMC47** [comment]
We should probably try to start pushing for this to be reviewed/merged asap so that it can be unleashed upon the users in development builds for a while.

As with the last two fixes; more bugs have been popping up from the timing change.

**Comment by phire** [comment]
For various reasons, I'm not actually expecting regressions from this one.

Though should be merged sooner rather than later.

**Comment by JMC47** [comment]
I'm not expecting regressions from this one, but, I'm curious if there are more still out there.  There are people saying we're getting invalid reads from Metroid Prime 1.02 yesterday during the escape sequence.

**Comment by mimimi085181** [comment]
Would an audio dma interrupt on hardware always take the same time? How hard would it be to write that hardware test?

@JMC47: Did those people correctly bisect to the merge of pr https://github.com/dolphin-emu/dolphin/pull/3800?

**Comment by phire** [comment]
@mimimi085181 not necessarily. My current theory is: 

There is a hardware timer triggering every few hundred (or thousand) cycles which clocks the next bit (or bits) into the the audio DAC. If the audio DMA registers have changed from disabled-enabled since the last timer, it latches the new values into the DMA and starts a transfer.

This would make the delay range from zero to a few hundred (or thousand) cycles depending on where in the phase this register is changed. 

I'm planning to write a hardware test to get the exact timings so I can fix [issue 9410](https://bugs.dolphin-emu.org/issues/9410). Early tests are indicating that there is enough time on hardware between this interrupt being triggered and data arriving at the DAC so you can re-load new values into the DMA without audible glitches.

**Comment by JMC47** [comment]
Please review/merge.  This is a big step toward 5.0!

**Comment by degasus** [comment]
Reviewed 1 of 1 files at r1.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/dolphin-emu/dolphin/3829)*

**Comment by degasus** [comment]
The 80 cycles was set by @booto in https://github.com/dolphin-emu/dolphin/pull/762
As we've changed the timing behavior of the JIT, his HW test may be wrong.

**Comment by phire** [comment]
booto didn't do a hardware test down the the cycle level. 

The major timing change in #762 was a change from the Audio DMA interrupt only being triggered at the end of a DMA transfer, to being triggered at the beginning of the transfer, including the first one when you enable the DMA.

**Comment by booto** [comment]
Just for posterity:
The test I did was super simple - just involved grabbing timebase register value as close to the start of the DMA transfer as possible, and grabs it again as first thing done in the interrupt handler, and then calcs the difference.

**Pull request merged**